tests/fallbacks/circuit_breaker/test_circuit_breaker.py

```diff
@@ -0,0 +1,277 @@
+            ) as circuit_breaker:
+                fail_function()
+
+            assert circuit_breaker.is_circuit_open
+
+    def test_should_raise_exception_when_circuit_is_open(
+        self,
+        max_failures_rule,
+        failure_cache_key
+    ):
+        circuit_cache_key = 'circuit_{}'.format(failure_cache_key)
+        cache.set(circuit_cache_key, True)
+
+        with pytest.raises(MyException):
+            with CircuitBreaker(
+                rule=max_failures_rule,
+                cache=cache,
+                failure_exception=MyException,
+                catch_exceptions=(ValueError,),
+            ) as circuit_breaker:
+                success_function()
+
+            assert circuit_breaker.is_circuit_open
+
+    def test_should_not_increment_fail_when_circuit_is_open(
+        self,
+        max_failures_rule,
+        failure_cache_key
+    ):
+        """
+        It should not increment fail count over the max failures limit, when
+        circuit breaker is open after a successful enter.
+        """
+        cache.set(failure_cache_key, 3)
+
+        with pytest.raises(MyException):
+            with CircuitBreaker(
+                rule=max_failures_rule,
+                cache=cache,
+                failure_exception=MyException,
+                catch_exceptions=(ValueError,),
+            ):
+                fail_function()
+
+        assert not cache.get(failure_cache_key)
+
+    def test_should_not_increment_request_when_circuit_is_open(
+        self,
+        percentage_failures_rule,
+        failure_cache_key,
+        request_cache_key
+    ):
+        """
+        It should not increment request count over the max failures limit, when
+        circuit breaker is open after a successful enter.
+        """
+        cache.set(failure_cache_key, 2)
+        cache.set(request_cache_key, 5)
+
+        with pytest.raises(MyException):
+            with CircuitBreaker(
+                rule=percentage_failures_rule,
+                cache=cache,
+                failure_exception=MyException,
+                catch_exceptions=(ValueError,),
+            ):
+                fail_function()
+
+        assert not cache.get(request_cache_key)
+
+    def test_should_delete_count_key_when_circuit_is_open(
+        self,
+        percentage_failures_rule,
+        failure_cache_key,
+        request_cache_key
+    ):
+        cache.set(failure_cache_key, 2)
+        cache.set(request_cache_key, 5)
+
+        with pytest.raises(MyException):
+            with CircuitBreaker(
+                rule=percentage_failures_rule,
+                cache=cache,
+                failure_exception=MyException,
+                catch_exceptions=(ValueError,),
+            ) as circuit_breaker:
+                fail_function()
+
+            assert circuit_breaker.is_circuit_open
+
+        assert cache.get(failure_cache_key) is None
+        assert cache.get(request_cache_key) is None
```

Comment:
As the `rule` implementations are independent and replaceable, you don't need to test the `Rule`s with the `CircuitBreaker`. I think would be better to implement fake rules that only have the possibilities allowed by the `CircuitBreaker`. This way you would not need to make the setup of percentage and max failures scenarios to test the CB.